res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+    } else {
+      if (TraktorS2MK2.effectFocusChooseModeActive[effectUnitGroup]) {
+        if (focusedEffect === buttonNumber) {
+          engine.setValue(effectUnitGroup, "focused_effect", 0);
+        } else {
+          engine.setValue(effectUnitGroup, "focused_effect", buttonNumber);
+        }
+        TraktorS2MK2.effectFocusChooseModeActive[effectUnitGroup] = false;
+      } else {
+        toggle();
+        TraktorS2MK2.effectButtonLongPressTimer[effectUnitGroup][buttonNumber] =
+          engine.beginTimer(TraktorS2MK2.longPressTimeout,
+            function() {
+              TraktorS2MK2.effectButtonIsLongPressed[effectUnitGroup][buttonNumber] = true;
+              TraktorS2MK2.effectButtonLongPressTimer[effectUnitGroup][buttonNumber] = 0;
+            },
+            true
+          );
+      }
+    }
+  } else {
+    engine.stopTimer(TraktorS2MK2.effectButtonLongPressTimer[effectUnitGroup][buttonNumber]);
+    TraktorS2MK2.effectButtonLongPressTimer[effectUnitGroup][buttonNumber] = 0;
+    if (TraktorS2MK2.effectButtonIsLongPressed[effectUnitGroup][buttonNumber]) {
+      toggle();
+    }
+    TraktorS2MK2.effectButtonIsLongPressed[effectUnitGroup][buttonNumber] = false;
+  }
+}
+
+TraktorS2MK2.topEncoder = function(field) {
+  // TODO: common-hid-packet-parser looks like it should do deltas, but I can't get them to work.
+  prev_pregain = TraktorS2MK2.previous_pregain[field.group];
+  TraktorS2MK2.previous_pregain[field.group] = field.value;
+  var delta = 0;
+  if (prev_pregain === 15 && field.value === 0) {
+    delta = 0.05;
+  } else if (prev_pregain === 0 && field.value === 15) {
+    delta = -0.05;
+  } else if (field.value > prev_pregain) {
+    delta = 0.05;
+  } else {
+    delta = -0.05;
+  }
+
+  if (TraktorS2MK2.shift_pressed[field.group]) {
+    var cur_pregain = engine.getValue(group, "pregain");
+    engine.setValue(group, "pregain", cur_pregain + delta);
+  } else {
+    var quickeffect_group = "[QuickEffectRack1_" + field.group + "]";
+    if (TraktorS2MK2.top_encoder_pressed[field.group]) {
+      script.triggerControl(quickeffect_group, delta > 0 ? "next_chain" : "prev_chain");
+    } else {
+      var cur_quickeffect = engine.getValue(quickeffect_group, "super1");
+      engine.setParameter(quickeffect_group, "super1", cur_quickeffect + delta / 1.5);
+    }
+  }
+}
+
+TraktorS2MK2.topEncoderPress = function(field) {
+  if (field.value > 0) {
+    TraktorS2MK2.top_encoder_pressed[field.group] = true;
+    if (TraktorS2MK2.shift_pressed[field.group]) {
+      script.triggerControl(field.group, "pregain_set_default");
+    } else {
+      script.triggerControl("[QuickEffectRack1_" + field.group + "]", "super1_set_default");
+    }
+  } else {
+    TraktorS2MK2.top_encoder_pressed[field.group] = false;
+  }
+}
+
+TraktorS2MK2.leftEncoder = function(field) {
+  // TODO: common-hid-packet-parser looks like it should do deltas, but I can't get them to work.
```

Comment:
are they used in any other mappings?

Reply:
Again, I think you wrote this comment :P

Reply:
Have a look just in case it's fixed itself since then? :)

Reply:
I couldn't get the library to work either and instead factored this code into a common encoder handling function for the controller in 6914d09